// tgdb/lib/pdu/impl/TransactionStatus.js
/**
 * Copyright 2016 TIBCO Software Inc. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except 
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

exports.TransactionStatus = {
            TGTransactionInvalid                   : -1,
            TGTransactionSuccess                   : 0,
            TGTransactionAlreadyInProgress         : 8001,
            TGTransactionClientDisconnected        : 8002,
            TGTransactionMalFormed                 : 8003,
            TGTransactionGeneralError              : 8004,
            TGTransactionVerificationError         : 8005,
            TGTransactionInBadState                : 8006,
            TGTransactionUniqueConstraintViolation : 8007,
            TGTransactionOptimisticLockFailed      : 8008,
            TGTransactionResourceExceeded          : 8009,
            TGCurrentThreadNotinTransaction        : 8010,
            TGTransactionUniqueIndexKeyAttributeNullError : 8011,
    	    fromStatus : function (status) {
            	switch(status) {
            		case -1 : return this.TGTransactionInvalid;
            		case 0 : return this.TGTransactionSuccess;
            		case 8001 : return this.TGTransactionAlreadyInProgress;
            		case 8002 : return this.TGTransactionClientDisconnected;
            		case 8003 : return this.TGTransactionMalFormed;
            		case 8004 : return this.TGTransactionGeneralError;
            		case 8005 : return this.TGTransactionVerificationError;
            		case 8006 : return this.TGTransactionInBadState;
            		case 8007 : return this.TGTransactionUniqueConstraintViolation;
            		case 8008 : return this.TGTransactionOptimisticLockFailed;
            		case 8009 : return this.TGTransactionResourceExceeded;
            		case 8010 : return this.TGCurrentThreadNotinTransaction;
            		case 8011 : return this.TGTransactionUniqueIndexKeyAttributeNullError;
    	    }
    	}
    };
