// tgdb/index.js
/**
 * Copyright 2016 TIBCO Software Inc. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except 
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

exports.TGConnectionFactory = require('./lib/connection/TGConnectionFactory');
exports.TGQueryOption       = require('./lib/query/TGQueryOption').TGQueryOption,
exports.TGLogManager        = require('./lib/log/TGLogManager');
exports.TGLogLevel          = require('./lib/log/TGLogger').TGLogLevel;
exports.TGEntityKind        = require('./lib/model/TGEntityKind').TGEntityKind;
exports.PrintUtility        = require('./lib/utils/PrintUtility');
exports.TGException         = require('./lib/exception/TGException').TGException;
exports.TGTransactionException                          = require('./lib/exception/TGTransactionException').TGTransactionException;
exports.TGTransactionAlreadyInProgressException         = require('./lib/exception/TGTransactionException').TGTransactionAlreadyInProgressException;
exports.TGTransactionMalFormedException                 = require('./lib/exception/TGTransactionException').TGTransactionMalFormedException;
exports.TGTransactionGeneralErrorException              = require('./lib/exception/TGTransactionException').TGTransactionGeneralErrorException;
exports.TGTransactionInBadStateException                = require('./lib/exception/TGTransactionException').TGTransactionInBadStateException;
exports.TGTransactionVerificationErrorException         = require('./lib/exception/TGTransactionException').TGTransactionVerificationErrorException;
exports.TGTransactionUniqueConstraintViolationException = require('./lib/exception/TGTransactionException').TGTransactionUniqueConstraintViolationException;
exports.TGTransactionOptimisticLockFailedException      = require('./lib/exception/TGTransactionException').TGTransactionOptimisticLockFailedException;
exports.TGTransactionResourceExceededException          = require('./lib/exception/TGTransactionException').TGTransactionResourceExceededException;
