// tgdb/lib/query/impl/TGQueryOptionImpl.js
/**
 * Copyright 2016 TIBCO Software Inc. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except 
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var TGLogManager              = require('../../log/TGLogManager'),
    TGLogLevel                = require('../../log/TGLogger').TGLogLevel

var QUERY_OPTION_FETCHSIZE = "fetchsize";
var QUERY_OPTION_BATCHSIZE = "batchsize";
var QUERY_OPTION_TRAVERSALDEPTH = "traversaldepth";
var QUERY_OPTION_EDGELIMIT = "edgelimit";
// Since 1.1
var QUERY_OPTION_SORTATTR = "sortattrname";
var QUERY_OPTION_SORTORDER = "sortorder"; // 0 - asc, 1 - dsc
var QUERY_OPTION_SORTLIMIT = "sortresultlimit";
var logger = TGLogManager.getLogger();

function TGQueryOptionImpl (mutable) {
    this._mutable = mutable;
    this._properties = {};
    this._properties[QUERY_OPTION_FETCHSIZE] = -1;
    this._properties[QUERY_OPTION_TRAVERSALDEPTH] = -1;
    this._properties[QUERY_OPTION_EDGELIMIT] = -1;
    this._edgeFilter = null;
    this._traversalCondition = null;
    this._endCondition = null;
}

TGQueryOptionImpl.prototype.setPrefetchSize = function(size) {
    if (!this._mutable) {
    	throw new Error("Can't modify a immutable Option");
    }
    if (size === 0) {
    	size = -1;
    }
    this._properties[QUERY_OPTION_FETCHSIZE] = size;
	logger.logDebug("[TGQueryOptionImpl::setPrefetchSize] %s = %d", QUERY_OPTION_FETCHSIZE, this._properties[QUERY_OPTION_FETCHSIZE]);

};

TGQueryOptionImpl.prototype.getPrefetchSize = function() {
	logger.logDebug("[TGQueryOptionImpl::getPrefetchSize] %s = %d", QUERY_OPTION_FETCHSIZE, this._properties[QUERY_OPTION_FETCHSIZE]);
    return this._properties[QUERY_OPTION_FETCHSIZE];
};

TGQueryOptionImpl.prototype.setBatchSize = function(size) {
    if (!this._mutable) {
    	throw new Error("Can't modify a immutable Option");
    }
    if (size === 0) {
    	size = -1;
    }
    this._properties[QUERY_OPTION_BATCHSIZE] = size;
	logger.logDebug("[TGQueryOptionImpl::setBatchSizeSize] %s = %d", QUERY_OPTION_BATCHSIZE, this._properties[QUERY_OPTION_BATCHSIZE]);

};

TGQueryOptionImpl.prototype.getBatchSize = function() {
	logger.logDebug("[TGQueryOptionImpl::getBatchSizeSize] %s = %d", QUERY_OPTION_BATCHSIZE, this._properties[QUERY_OPTION_BATCHSIZE]);
    return this._properties[QUERY_OPTION_BATCHSIZE];
};

TGQueryOptionImpl.prototype.setTraversalDepth = function(depth) {
    if (!this._mutable) {
        throw new Error("Can't modify a immutable Option");
    }
    
    if (depth === 0) {
        depth = -1;
    }
    this._properties[QUERY_OPTION_TRAVERSALDEPTH] = depth;
	logger.logInfo("[TGQueryOptionImpl::setTraversalDepth] %s = %d", QUERY_OPTION_TRAVERSALDEPTH, this._properties[QUERY_OPTION_TRAVERSALDEPTH]);
};

TGQueryOptionImpl.prototype.getTraversalDepth = function() {
	logger.logDebug("[TGQueryOptionImpl::getTraversalDepth] %s = %d", QUERY_OPTION_TRAVERSALDEPTH, this._properties[QUERY_OPTION_TRAVERSALDEPTH]);
    return this._properties[QUERY_OPTION_TRAVERSALDEPTH];
};

TGQueryOptionImpl.prototype.setEdgeLimit = function(limit) {
    if (!this._mutable) {
    	throw new Error("Can't modify a immutable Option");
    }
    if (limit === 0) {
    	limit = -1;
    }
    this._properties[QUERY_OPTION_EDGELIMIT] = limit;
	logger.logDebug("[TGQueryOptionImpl::setEdgeLimit] %s = %d", QUERY_OPTION_EDGELIMIT, this._properties[QUERY_OPTION_EDGELIMIT]);
};

TGQueryOptionImpl.prototype.getEdgeLimit = function() {
	logger.logDebug("[TGQueryOptionImpl::getEdgeLimit] %s = %d", QUERY_OPTION_EDGELIMIT, this._properties[QUERY_OPTION_EDGELIMIT]);
    return this._properties[QUERY_OPTION_EDGELIMIT];
};

// Since 1.1
TGQueryOptionImpl.prototype.setSortAttrName = function(name) {
    if (!this._mutable) { 
    	throw new TGException("Can't modify a immutable Option");
    }
    if (!name||0===name.length()) {
    	throw new TGException("Invalid attribute name");
    }
    this._properties[QUERY_OPTION_SORTATTR] = name;
	logger.logDebug("[TGQueryOptionImpl::setSortAttrName] %s = %d", QUERY_OPTION_SORTATTR, this._properties[QUERY_OPTION_SORTATTR]);
}

//Since 1.1
TGQueryOptionImpl.prototype.getSortAttrName = function () {
    return this._properties[QUERY_OPTION_SORTATTR];
	logger.logDebug("[TGQueryOptionImpl::getSortAttrName] %s = %d", QUERY_OPTION_SORTATTR, this._properties[QUERY_OPTION_SORTATTR]);
}

//Since 1.1
TGQueryOptionImpl.prototype.setSortOrderDsc = function (isDsc) {
    if (!this._mutable) {
    	throw new TGException("Can't modify a immutable Option");
    }
    if (isDsc) {
    	this._properties[QUERY_OPTION_SORTORDER] = "1";
    } else {
    	this._properties[QUERY_OPTION_SORTORDER] = "0";
    }
	logger.logDebug("[TGQueryOptionImpl::setSortOrderDsc] %s = %d", QUERY_OPTION_SORTORDER, this._properties[QUERY_OPTION_SORTORDER]);
}

//Since 1.1
TGQueryOptionImpl.prototype.isSortOrderDsc = function () {
	logger.logDebug("[TGQueryOptionImpl::getSortOrderDsc] %s = %d", QUERY_OPTION_SORTORDER, this._properties[QUERY_OPTION_SORTORDER]);
    return this._properties[QUERY_OPTION_SORTORDER];
}

//Since 1.1
TGQueryOptionImpl.prototype.setSortResultLimit = function (limit) {
    if (!this._mutable) {
    	throw new TGException("Can't modify a immutable Option");
    }
    if (limit <= 0) {
    	throw new TGException("Invalid sort limit");
    }
    this._properties[QUERY_OPTION_SORTLIMIT] = limit+'';
	logger.logDebug("[TGQueryOptionImpl::setSortResultLimit] %s = %d", QUERY_OPTION_SORTLIMIT, this._properties[QUERY_OPTION_SORTLIMIT]);
}

//Since 1.1
TGQueryOptionImpl.prototype.getSortResultLimit = function () {
	logger.logDebug("[TGQueryOptionImpl::getSortResultLimit] %s = %d", QUERY_OPTION_SORTLIMIT, this._properties[QUERY_OPTION_SORTLIMIT]);
    return this._properties[QUERY_OPTION_SORTLIMIT];
}

//Since 1.1
TGQueryOptionImpl.prototype.setEdgeFilter = function (edgeFilter) {
    if (!this._mutable) {
    	throw new TGException("Can't modify a immutable Option");
    }
    this._edgeFilter = edgeFilter;
	logger.logDebug("[TGQueryOptionImpl::setEdgeFilter] _edgeFilter = " + this._edgeFilter);
}

//Since 1.1
TGQueryOptionImpl.prototype.getEdgeFilter = function () {
	logger.logDebug("[TGQueryOptionImpl::getEdgeFilter] _edgeFilter = " + this._edgeFilter);
    return this._edgeFilter;
}

//Since 1.1
TGQueryOptionImpl.prototype.setTraversalCondition = function (traversalCondition) {
    if (!this._mutable) {
    	throw new TGException("Can't modify a immutable Option");
    }
    this._traversalCondition = traversalCondition;
	logger.logDebug("[TGQueryOptionImpl::setTraversalCondition] %s = %s", "_traversalCondition", this._traversalCondition);
}

//Since 1.1
TGQueryOptionImpl.prototype.getTraversalCondition = function () {
	logger.logDebug("[TGQueryOptionImpl::getTraversalCondition] %s = %s", "_traversalCondition", this._traversalCondition);
    return this._traversalCondition;
}

//Since 1.1
TGQueryOptionImpl.prototype.setEndCondition = function (endCondition) {
    if (!this._mutable) {
    	throw new TGException("Can't modify a immutable Option");
    }
    this._endCondition = endCondition;
	logger.logDebug("[TGQueryOptionImpl::setEndCondition] %s = %s", "_endCondition", this._endCondition);
}

//Since 1.1
TGQueryOptionImpl.prototype.getEndCondition = function () {
	logger.logDebug("[TGQueryOptionImpl::getEndCondition] %s = %s", "_endCondition", this._endCondition);
    return this._endCondition;
}

exports.TGQueryOptionImpl = TGQueryOptionImpl;

function test() {
	var option = new TGQueryOptionImpl(false);
}

//test();
