// tgdb/lib/model/TGEdgeType.js
/**
 * Copyright 2016 TIBCO Software Inc. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except 
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var TGEntityType        = require('./TGEntityType'),
	TGEdgeDirectionType = require('./TGEdgeDirectionType').TGEdgeDirectionType,
	TGSystemObject      = require('./TGSystemObject').TGSystemObject,
    TGAbstractEntity    = require('./TGAbstractEntity').TGAbstractEntity,
    util                = require('util');

function TGEdgeType(graphMetadata, name, directionType, parentType) {
	TGEdgeType.super_.call(this);
//    this._name           = name;
    this._directionType  = directionType;
    this._parentType     = parentType;
    
    // From 1.1
    this._fromTypeId = null;
    this._toTypeId = null;
    this._fromNodeType = null;
    this._toNodeType = null;
}

util.inherits(TGEdgeType, TGEntityType);

// From 1.1
TGEdgeType.prototype.getFromId= function() {
    return this._fromTypeId;
};

// From 1.1
TGEdgeType.prototype.getToId= function() {
    return this._toTypeId;
};

// From 1.1
TGEdgeType.prototype.updateMetadata= function(gmd) {
    this._fromNodeType = gmd.getNodeType(this._fromTypeId);
    this._toNodeType = gmd.getNodeType(this._toTypeId);
};

TGEdgeType.prototype.getSystemType = function() {
    return TGSystemObject.TGSystemType.EdgeType;
};

TGEdgeType.prototype.getDirectionType = function() {
    return this._directionType;
};

// From 1.1
TGEdgeType.prototype.getFromNodeType = function() {
    return this._fromNodeType;
};

// From 1.1
TGEdgeType.prototype.getToNodeType = function() {
    return this._toNodeType;
};

// From 1.1
TGEdgeType.prototype.readExternal = function(inputStream) {
	this.readAttributeDescriptors(inputStream);

    this._fromTypeId = inputStream.readInt();
    this._toTypeId = inputStream.readInt();
    var dir = inputStream.readByte();
    switch (dir) {
    	case TGEdgeDirectionType.UNDIRECTED.ordinal :
    		this._directionType = TGEdgeDirectionType.UNDIRECTED;
    		break;
    	case TGEdgeDirectionType.DIRECTED.ordinal :
    		this._directionType = TGEdgeDirectionType.DIRECTED;
    		break;
    	case TGEdgeDirectionType.BIDIRECTIONAL.ordinal :
    		this._directionType = TGEdgeDirectionType.BIDIRECTIONAL;
    		break;
    }
};

module.exports = TGEdgeType;
