// tgdb-core/dist/utils/parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseData = exports.parseTable = exports.parseIndex = void 0;
function parseIndex(str) {
    const lines = str.split("\n");
    return {
        header: parseHeader(lines[0]),
        table: parseTable(lines.slice(1)),
    };
}
exports.parseIndex = parseIndex;
function parseTable(tables) {
    const list = {};
    for (const table of tables) {
        const [tableName, msgId] = table.split(" ");
        const messageId = parseInt(msgId);
        if (!isNaN(messageId))
            list[tableName] = messageId;
    }
    return list;
}
exports.parseTable = parseTable;
function parseData(str) {
    const lines = str.split("\n");
    if (lines.length > 1)
        lines.pop();
    return {
        header: parseHeader(lines[0]),
        dbName: lines[1].split(" ")[0],
        dbMsgId: parseInt(lines[1].split(" ")[1]),
        value: lines.slice(2).join("\n"),
    };
}
exports.parseData = parseData;
function parseNumber(str) {
    return isNaN(parseInt(str)) ? null : parseInt(str);
}
function parseHeader(str) {
    const headers = str.split(" ");
    return {
        page_id: headers[0],
        page_index: parseInt(headers[1]),
        prev_msg_id: parseNumber(headers[2]),
        next_msg_id: parseNumber(headers[3]),
    };
}
